File: python/board.py
from __future__ import annotations
import sys
from enum import Enum
from typing import Union, List

class Cell:
    def __init__(self, 
                 clear: bool=True, 
                 value: Union[int, None]=0, 
                 revealed: bool=False,
                 flagged: bool=False) -> None:
        self.clear = clear
        self.value = value
        self.revealed = revealed
        self.flagged = flagged
    
    def incr_value(self) -> None:
        if not self.clear:
            raise Exception('cannot call incr_value on a non-clear cell')
        self.value += 1

    def reveal_cell(self) -> Union[int, None]:
        self.revealed = True
        if not self.clear:
            return 1

    def switch_flag(self) -> None:
        self.flagged = not self.flagged

    # checks if a cell is either revealed/flagged or is a mine
    def revealed_mine_or_flagged(self) -> bool:
        return self.revealed or self.flagged or not self.clear
    
    def __str__(self) -> str:
        return f'clear {self.clear}, value {self.value}, revealed {self.revealed}, flagged {self.flagged}'


class Position:
    def __init__(self, row: int, col: int) -> None:
        self.row = row
        self.col = col

    def __str__(self) -> str:
        return f'row {self.row}, col {self.col}'
    
    # def __eq__(self, other: Position) -> bool:
    #     return self.row == other.row and self.col == other.col
    
    # def __hash__(self) -> int:
    #     return hash((self.row, self.col))

class Board:
    def __init__(self, height: int, width: int) -> None:
        self.height = height
        self.width = width
        self.cells = [[Cell() for _ in range(width)] for _ in range(height)]

    def get_cell(self, pos: Position) -> Cell:
        i, j = pos.row, pos.col
        if i >= self.height or j >= self.width:
            raise Exception(f'position ({i}, {j}) is not on the board')
        return self.cells[i][j]
    
    def get_row(self, i: int) -> List[Cell]:
        if i >= self.height:
            raise Exception(f'row {i} is not on the board')
        return self.cells[i]
    
    def set_cell(self, pos: Position, cell: Cell) -> None:
        i, j = pos.row, pos.col
        if i >= self.height or j >= self.width:
            raise Exception(f'position ({i}, {j}) is not on the board')
        self.cells[i][j] = cell

    def __str__(self) -> str:
        res = ''
        for i in range(self.height):
            for j in range(self.width):
                current_cell = self.get_cell(Position(i, j))
                res += str(current_cell.value) if current_cell.clear else '*'
            res += '\n'
        return res
    
    def make_mine(self, pos: Position) -> None:
        i, j = pos.row, pos.col
        self.set_cell(Position(i, j), Cell(False, None))

    def on_board(self, pos: Position) -> bool:
        return 0 <= pos.row < self.height and 0 <= pos.col < self.width
    
    def get_adjacents(self, pos: Position) -> List[Position]:
        directions = (-1, 0, 1)
        res = []
        for dy in directions:
            for dx in directions:
                if dx == dy == 0: 
                    continue
                new_pos = Position(pos.row+dy, pos.col+dx)
                if not self.on_board(new_pos):
                    continue
                res.append(new_pos)
        return res
    
    def show_board(self) -> None:
        res = ''
        for i in range(self.height):
            for j in range(self.width):
                current_cell = self.get_cell(Position(i, j))
                res += str(current_cell.value) if current_cell.clear else '*'
            res += '\n'
        print(res)

File: c/board.c
#include "board.h"

position new_position(int row, int col) {
    position new;
    new.row = row;
    new.col = col;
    return new;
}

cell new_cell(int value, bool clear, bool revealed, bool flagged) {
    cell new;
    new.value = value;
    new.clear = clear;
    new.revealed = revealed;
    new.flagged = flagged;
    new.detonated = false;
    return new;
}

void incr_cell_value(cell *c) {
    c->value++;
}

int reveal_cell(cell *c) {
    c->revealed = true;
    return c->clear ? 0 : 1;
}

void switch_cell_flag(cell *c) {
    c->flagged = !c->flagged;
}

bool nonzero_mine_or_flagged(cell *c) {
    return !c->clear || c->value || c->flagged;
}

board *new_board(int height, int width) {
    board *b = (board*)malloc(sizeof(board));
    b->height = height;
    b->width = width;
    cell **matrix = (cell**)malloc(height * sizeof(cell*));
    for (int i = 0; i < height; i++) {
        matrix[i] = (cell*)malloc(width * sizeof(cell));
        for (int j = 0; j < width; j++) {
            matrix[i][j] = new_cell(0, true, false, false);
        }
    }
    b->cells = matrix;
    return b;
}

void free_board(board *b) {
    for (int i = 0; i < b->height; i++) {
        free(b->cells[i]);
    }
    free(b);
}

cell *get_cell(board *b, position pos) {
    return &b->cells[pos.row][pos.col];
}

cell **get_row(board *b, int i) {
    return &b->cells[i];
}

void set_cell(board *b, position pos, cell c) {
    cell *tmp = get_cell(b, pos);
    tmp->value = c.value;
    tmp->clear = c.clear;
    tmp->flagged = c.flagged;
    tmp->revealed = c.revealed;
    tmp->detonated = c.detonated;
}

void make_mine(board *b, position pos) {
    cell mine = {-1, false, false, false, false};
    set_cell(b, pos, mine);
}

bool on_board(board *b, position pos) {
    return 0 <= pos.row && pos.row < b->height && 0 <= pos.col && pos.col < b->width;
}

pos_ll *prepend(position to_prepend, pos_ll *head) {
    pos_ll *res = (pos_ll*)malloc(sizeof(pos_ll));
    res->curr = to_prepend;
    res->next = head;
    return res;
}

pos_ll *get_adjacents(board *b, position pos) {
    int directions[] = {-1, 0, 1};
    pos_ll *res = NULL;
    position new_pos;
    int dx, dy;
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            dx = directions[i], dy = directions[j];
            if (dx == dy && dy == 0) continue;
            new_pos = new_position(pos.row + dy, pos.col + dx);
            if (!on_board(b, new_pos)) continue;
            res = prepend(new_pos, res);
        }
    }
    return res;
}

void free_posll(pos_ll *head) {
    pos_ll *tmp;
    while (tmp = head) {
        head = head->next;
        free(tmp);
    }
}

void show_board(board *b) {
    position pos;
    cell *current_cell;
    for (int i = 0; i < b->height; i++) {
        for (int j = 0; j < b->width; j++) {
            pos = new_position(i, j);
            current_cell = get_cell(b, pos);
            printf("%c", current_cell->clear ? current_cell->value+'0' : '*');
        }
        printf("\n");
    }
}
